sched/semaphore/sem_rw.c

```diff
@@ -23,6 +23,20 @@
  ****************************************************************************/
 
 #include <nuttx/rwsem.h>
+#include <nuttx/irq.h>
+#include <assert.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#undef spin_lock_irqsave
+#undef spin_unlock_irqrestore
+
+#if !defined(CONFIG_SPINLOCK) || !defined(CONFIG_SMP)
+  #define spin_lock_irqsave(lock) up_irq_save()
```

Comment:
let's always define protected to avoid the redefinition like this.

Reply:
Sorry, but what is defined protected?
Could you give me an example?

Reply:
```
#if defined(CONFIG_SMP) || defined(CONFIG_SPINLOCK)
irqstate_t spin_lock_irqsave(FAR spinlock_t *lock);
#else
#  define spin_lock_irqsave(l) up_irq_save()
#endif
```

sim:posix_test or sim:zipfs are going build fail.
I have not check reason detail.

Reply:
cherry-pick and build log:
d2e466f7a744b6e52a16b0a277ace1cf57cc5bd8
https://github.com/apache/nuttx/actions/runs/8295556267/job/22703014434?pr=11769

Reply:
let's add `ifdef CONFIG_SPINLOCK` to g_sigaction_spin or remove it from rw_semaphore_t::protected
````
Error: signal/sig_action.c:47:19: error: 'g_sigaction_spin' defined but not used [-Werror=unused-variable]
   47 | static spinlock_t g_sigaction_spin;
      |                   ^~~~~~~~~~~~~~~~
```